src/nvim/rbuffer.c

```diff
@@ -34,17 +34,20 @@ void rbuffer_free(RBuffer *buf)
   xfree(buf);
 }
 
-size_t rbuffer_size(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_size(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
 {
   return buf->size;
 }
 
-size_t rbuffer_capacity(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_capacity(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
 {
   return (size_t)(buf->end_ptr - buf->start_ptr);
 }
 
-size_t rbuffer_space(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_space(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
```

Comment:
Thanks for mentioning that. I didn't know that pure is not allowed to be used if the parameter is used by multiple threads, that makes it extremely difficult to ever know if a function can safely be marked pure. @ZyX-I can you confirm?